// context.dev/resources/brand.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../core/resource.mjs";
export class Brand extends APIResource {
    /**
     * Retrieve logos, backdrops, colors, industry, description, and more from any
     * domain
     */
    retrieve(query, options) {
        return this._client.get('/brand/retrieve', { query, ...options });
    }
    /**
     * Endpoint specially designed for platforms that want to identify transaction data
     * by the transaction title.
     */
    identifyFromTransaction(query, options) {
        return this._client.get('/brand/transaction_identifier', { query, ...options });
    }
    /**
     * Retrieve brand information using an email address while detecting disposable and
     * free email addresses. Disposable and free email addresses (like gmail.com,
     * yahoo.com) will throw a 422 error.
     */
    retrieveByEmail(query, options) {
        return this._client.get('/brand/retrieve-by-email', { query, ...options });
    }
    /**
     * Retrieve brand information using an ISIN (International Securities
     * Identification Number).
     */
    retrieveByIsin(query, options) {
        return this._client.get('/brand/retrieve-by-isin', { query, ...options });
    }
    /**
     * Retrieve brand information using a company name.
     */
    retrieveByName(query, options) {
        return this._client.get('/brand/retrieve-by-name', { query, ...options });
    }
    /**
     * Retrieve brand information using a stock ticker symbol.
     */
    retrieveByTicker(query, options) {
        return this._client.get('/brand/retrieve-by-ticker', { query, ...options });
    }
    /**
     * Returns a simplified version of brand data containing only essential
     * information: domain, title, colors, logos, and backdrops. Optimized for faster
     * responses and reduced data transfer.
     */
    retrieveSimplified(query, options) {
        return this._client.get('/brand/retrieve-simplified', { query, ...options });
    }
}
