// context.dev/resources/web.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../core/resource.mjs";
export class Web extends APIResource {
    /**
     * Crawl a website, use the provided JSON Schema and instructions to prioritize
     * relevant internal links, and extract structured data from the selected pages.
     *
     * @example
     * ```ts
     * const response = await client.web.extract({
     *   schema: {
     *     type: 'bar',
     *     properties: 'bar',
     *     required: 'bar',
     *     additionalProperties: 'bar',
     *   },
     *   url: 'https://example.com',
     * });
     * ```
     */
    extract(body, options) {
        return this._client.post('/web/extract', { body, ...options });
    }
    /**
     * Analyze a company's landing page and web search evidence to return direct
     * competitors for the same product or market.
     *
     * @example
     * ```ts
     * const response = await client.web.extractCompetitors({
     *   domain: 'xxx',
     * });
     * ```
     */
    extractCompetitors(query, options) {
        return this._client.get('/web/competitors', { query, ...options });
    }
    /**
     * Scrape font information from a website including font families, usage
     * statistics, fallbacks, and element/word counts.
     *
     * @example
     * ```ts
     * const response = await client.web.extractFonts();
     * ```
     */
    extractFonts(query = {}, options) {
        return this._client.get('/web/fonts', { query, ...options });
    }
    /**
     * Extract a comprehensive design system from a website including colors,
     * typography, spacing, shadows, and UI components.
     *
     * @example
     * ```ts
     * const response = await client.web.extractStyleguide();
     * ```
     */
    extractStyleguide(query = {}, options) {
        return this._client.get('/web/styleguide', { query, ...options });
    }
    /**
     * Capture a screenshot of a website.
     *
     * @example
     * ```ts
     * const response = await client.web.screenshot();
     * ```
     */
    screenshot(query = {}, options) {
        return this._client.get('/web/screenshot', { query, ...options });
    }
    /**
     * Search the web and optionally scrape each result to Markdown in one round-trip.
     *
     * @example
     * ```ts
     * const response = await client.web.search({ query: 'x' });
     * ```
     */
    search(body, options) {
        return this._client.post('/web/search', { body, ...options });
    }
    /**
     * Performs a crawl starting from a given URL, extracts page content as Markdown,
     * and returns results for all crawled pages.
     *
     * @example
     * ```ts
     * const response = await client.web.webCrawlMd({
     *   url: 'https://example.com',
     * });
     * ```
     */
    webCrawlMd(body, options) {
        return this._client.post('/web/crawl', { body, ...options });
    }
    /**
     * Scrapes the given URL and returns the raw HTML content of the page.
     *
     * @example
     * ```ts
     * const response = await client.web.webScrapeHTML({
     *   url: 'https://example.com',
     * });
     * ```
     */
    webScrapeHTML(query, options) {
        return this._client.get('/web/scrape/html', { query, ...options });
    }
    /**
     * Extract image assets from a web page, including standard URLs, inline SVGs, data
     * URIs, responsive image sources, metadata, CSS backgrounds, video posters, and
     * embeds. The base request costs 1 credit. When enrichment is enabled, the entire
     * call costs 5 credits.
     *
     * @example
     * ```ts
     * const response = await client.web.webScrapeImages({
     *   url: 'https://example.com',
     * });
     * ```
     */
    webScrapeImages(query, options) {
        return this._client.get('/web/scrape/images', { query, ...options });
    }
    /**
     * Scrapes the given URL into LLM usable Markdown.
     *
     * @example
     * ```ts
     * const response = await client.web.webScrapeMd({
     *   url: 'https://example.com',
     * });
     * ```
     */
    webScrapeMd(query, options) {
        return this._client.get('/web/scrape/markdown', { query, ...options });
    }
    /**
     * Crawl an entire website's sitemap and return all discovered page URLs.
     *
     * @example
     * ```ts
     * const response = await client.web.webScrapeSitemap({
     *   domain: 'domain',
     * });
     * ```
     */
    webScrapeSitemap(query, options) {
        return this._client.get('/web/scrape/sitemap', { query, ...options });
    }
}
